src/models/state/mempool.rs

```diff
@@ -226,31 +239,55 @@ impl Mempool {
+    /// in the mempool are checked if they constitute a matching pair with the
+    /// 1st selected transaction. If no such pair can be found, the transaction
+    /// with the second-highest fee density is checked against the rest of the
+    /// transactions and so on.
+    ///
+    /// Returns `None` if no such pair exists in the mempool, or if such a pair
+    /// exists but its combined number of inputs exceeds the mempool's policy
+    /// for maximum number of inputs.
     pub(crate) fn most_dense_single_proof_pair(
         &self,
     ) -> Option<([(&TransactionKernel, &Proof); 2], TransactionOrigin)> {
-        let mut ret = vec![];
-        let mut own_tx = false;
-        for (txid, _fee_density) in self.get_sorted_iter() {
-            let candidate = self.tx_dictionary.get(&txid).unwrap();
-            if let TransactionProof::SingleProof(proof) = &candidate.transaction.proof {
-                ret.push((&candidate.transaction.kernel, proof));
-                own_tx = own_tx || candidate.origin.is_own();
-            }
-
-            let origin = match own_tx {
-                true => TransactionOrigin::Own,
-                false => TransactionOrigin::Foreign,
-            };
+        /// Loop from `i`th densest transaction in the mempool to the
+        /// least dense in order to find a pair whose total number of inputs
+        /// does not exceed the maximum value.
+        fn most_dense_pair_skip_first_n(
+            mempool: &Mempool,
+            i: usize,
+        ) -> Option<([(&TransactionKernel, &Proof); 2], TransactionOrigin)> {
+            let mut ret = vec![];
+            let mut own_tx = false;
+            let mut remaining_num_inputs = mempool.max_num_inputs_per_transaction as i128;
+            for (txid, _fee_density) in mempool.get_sorted_iter().skip(i) {
+                let candidate = mempool.tx_dictionary.get(&txid).unwrap();
+                let candidate_num_inputs = candidate.transaction.kernel.inputs.len() as i128;
+                if let TransactionProof::SingleProof(proof) = &candidate.transaction.proof {
+                    if remaining_num_inputs >= candidate_num_inputs {
+                        ret.push((&candidate.transaction.kernel, proof));
+                        own_tx = own_tx || candidate.origin.is_own();
+                        remaining_num_inputs -= candidate_num_inputs;
+                    }
+                }
 
-            if ret.len() == 2 {
-                return Some((ret.try_into().unwrap(), origin));
+                if ret.len() == 2 {
+                    let origin = match own_tx {
+                        true => TransactionOrigin::Own,
+                        false => TransactionOrigin::Foreign,
+                    };
+                    return Some((ret.try_into().unwrap(), origin));
+                }
             }
+
+            None
         }
 
-        None
+        (0..self.len() - 1).find_map(|i| most_dense_pair_skip_first_n(self, i))
```

Comment:
I think the expression `self.len() - 1` can overflow. Probably not good.

Reply:
Just shows that these getters need to be tested on an empty mempool. That would have caught this problem by producing a never-ending test case.